spring-integration-jdbc/src/main/java/org/springframework/integration/jdbc/metadata/JdbcMetadataStore.java

```diff
@@ -0,0 +1,198 @@
+					return this.jdbcTemplate.queryForObject(getQuery(Query.GET_VALUE), String.class, key);
+				}
+				catch (EmptyResultDataAccessException e) {
+					//somebody deleted it between calls. try to insert again (go to beginning of while loop)
+				}
+			}
+		}
+	}
+
+	@Override
+	public boolean replace(String key, String oldValue, String newValue) {
+		Assert.notNull(key, "'key' cannot be null");
+		Assert.notNull(oldValue, "'oldValue' cannot be null");
+		Assert.notNull(newValue, "'newValue' cannot be null");
+		int affectedRows = this.jdbcTemplate.update(getQuery(Query.REPLACE_VALUE), ps -> {
+			ps.setString(1, newValue);
+			ps.setString(2, key);
+			ps.setString(3, oldValue);
+		});
+		return affectedRows > 0;
+	}
+
+	@Override
+	@Transactional(propagation = Propagation.REQUIRES_NEW)
+	public void put(String key, String value) {
+		Assert.notNull(key, "'key' cannot be null");
+		Assert.notNull(value, "'value' cannot be null");
+		while (true) {
+			//try to insert if does not exist, if exists we will try to update it
+			int affectedRows = this.jdbcTemplate.update(getQuery(Query.PUT_IF_ABSENT_VALUE), ps -> {
+				ps.setString(1, key);
+				ps.setString(2, value);
+				ps.setString(3, key);
+			});
+			if (affectedRows == 0) {
+				//since value is not inserted, means it is already present
+				try {
+					//lock row for updating
+					this.jdbcTemplate.queryForObject(getQuery(Query.GET_VALUE_FOR_UPDATE), String.class, key);
+				}
+				catch (EmptyResultDataAccessException e) {
+					//if there are no rows with this key, somebody deleted it in between two calls
+					continue;	//try to insert again from beginning
+				}
+				//lock successful, so - replace
+				this.jdbcTemplate.update(getQuery(Query.REPLACE_VALUE_BY_KEY), ps -> {
+					ps.setString(1, value);
+					ps.setString(2, key);
+				});
+			}
+			return;
+		}
+	}
+
+	@Override
+	public String get(String key) {
+		Assert.notNull(key, "'key' cannot be null");
+		try {
+			return this.jdbcTemplate.queryForObject(getQuery(Query.GET_VALUE), String.class, key);
+		}
+		catch (EmptyResultDataAccessException e) {
+			//if there are no rows with this key, return null
+			return null;
+		}
+	}
+
+	@Override
+	@Transactional(propagation = Propagation.REQUIRES_NEW)
+	public String remove(String key) {
+		Assert.notNull(key, "'key' cannot be null");
+		String oldValue;
+		try {
+			//select old value and lock row for removal
+			oldValue = this.jdbcTemplate.queryForObject(getQuery(Query.GET_VALUE_FOR_UPDATE), String.class, key);
+		}
+		catch (EmptyResultDataAccessException e) {
+			//key is not present, so no need to delete it
+			return null;
+		}
+		//delete row and return old value
+		int updated = this.jdbcTemplate.update(getQuery(Query.REMOVE_VALUE), key);
+		if (updated != 0) {
+			return oldValue;
+		}
+		return null;
+	}
+
+	private String getQuery(Query base) {
+		return this.queryCache.computeIfAbsent(base, b ->
+				StringUtils.replace(b.getSql(), "%PREFIX%", DEFAULT_TABLE_PREFIX));
```

Comment:
If there is a `DEFAULT_TABLE_PREFIX`, then there should be a configurable property on the matter:
```
/**
	 * Specify the prefix for target data base table used from queries.
	 * @param prefix the prefix to set (default INT_).
	 */
	public void setPrefix(String prefix) {
		this.prefix = prefix;
	}
```
Also, I think the `format` solution in the `DefaultLockRepository` is the best from the performance perspective.